### ISSUE TITLE
[BERT/Pytorch] Standard API call of checkpoint saving and loading not work, but solutions in Pytorch BERT examples work fine?

### ISSUE THREAD

**Issue by mxmxmxmx** [NVIDIA/DeepLearningExamples#397]
**Model/Framework(s)** 
BERT/PyTorch on muti GPUs with gradient accumulation, apex DDP, amp

training the model for many epochs and trying to save checkpoint and restart the run from where its left. 

**Describe the bug**

I was following the instructions below, save 
_'amp': amp.state_dict()_ 
and load it back using 
_amp.load_state_dict(checkpoint['amp'])._ 
The new run loaded from previous checkpoint goes back to a high training loss, and saved checkpoint is not working.
https://nvidia.github.io/apex/amp.html#checkpointing

Then I followed Nvidia Pytorch BERT examples below, save
 _'master params': list(amp.master_params(optimizer))_ 
and load it as 
           _optimizer._lazy_init_maybe_master_weights()
            optimizer._amp_stash.lazy_init_called = True
            optimizer.load_state_dict(checkpoint['optimizer'])
            for param, saved_param in zip(amp.master_params(optimizer), checkpoint['master params']):
                param.data.copy_(saved_param.data)_
https://github.com/NVIDIA/DeepLearningExamples/blob/0845aaa90187e0b79d244648deb2c9656bd6fb66/PyTorch/LanguageModeling/BERT/run_pretraining.py#L332

This works for me.

My question is:
why standard API checkpoint commands not work? Isn't calls in BERT example like "_optimizer._amp_stash.lazy_init_called_ " been taken care of in the official API "_amp.load_state_dict(checkpoint['amp'])_"

**Comment by sharathts** [comment]
We need the additional complexity in code with checkpoints since we are running with AMP optimization level O2. The way checkpoints are handled under the hood here is different compared to the default optimization level O1( with O1, this complexity isn't needed). O2 gives > 20% performance boost over O1.

**Comment by sharathts** [comment]
Closing issue. Feel free to reopen if you have more questions.
